spec/src/main/asciidoc/method-query.asciidoc

```diff
@@ -104,35 +104,52 @@ Query methods allow developers to create database queries using method naming co
 
 [source,bnf]
 ----
-<query-method> ::= <subject> <predicate> [<order-clause>]
-<subject> ::= (<action> | "find" <find-expression>) [<ignored-text>] "By"
-<action> ::= "find" | "delete" | "update" | "count" | "exists"
-<find-expression> ::= "First" [<positive-integer>]
-<predicate> ::= <condition> { ("And" | "Or") <condition> }
-<condition> ::= <property> ["IgnoreCase"] ["Not"] [<operator>]
-<operator> ::= "Contains" | "EndsWith" | "StartsWith" | "LessThan"| "LessThanEqual" | "GreaterThan" | "GreaterThanEqual" | "Between" | "Like" | "In" | "Null" | "True" | "False"
-<property> ::= <identifier> | <identifier> "_" <property>
-<identifier> ::= <word>
-<positive-integer> ::= <digit> { <digit> }
-<order-clause> ::= "OrderBy" { <order-item> } ( <order-item> | <property> )
-<order-item> ::= <property> ("Asc" | "Desc")
+query : find | action
+find : "find" specification order?
+action : ("delete" | "update" | "count" | "exists") restriction
+specification : "All" ignoredText? | limit? ignoredText? "By" predicate
+restriction : "All" ignoredText? | ignoredText? "By" predicate
```

Comment:
The following is simpler, more flexible, and does not require adding `All` as a keyword. Instead, `All` becomes part of the ignored text. The restriction clause beginning with "By" becomes optional. When omitted it means no restrictions, which for prefixes other than `exists` has the meaning of all, so you can have `count()` or `countAll()` but not `countBy()` because the latter needs at least one condition, `countByNameLike(pattern)`. When using the `exists` prefix, the repository writer can name it more meaningfully, such as `existsAny()` or `existsOne()`, to reflect what it actually does. This avoids having the confusing `existsAll()` method name be something that is recommended by the grammar.

```suggestion
find : "find" limit? ignoredText? restriction? order?
action : ("delete" | "update" | "count" | "exists") ignoredText? restriction?
restriction : "By" predicate
```

Reply:
The main reason I did not do it that way was I was a bit uncomfortable with allowing just plain `List<E> find()` and `void delete()` without any other qualification. But I guess it's OK. It simplifies the grammar. And this is an extension anyway.

But, wait, another question is: why is `update` an allowed action in this grammar?? I don't think we actually have update queries.

Reply:
I have removed `update`, and also updated the text to reflect Nathan's change to the grammar.

Reply:
> But, wait, another question is: why is `update` an allowed action in this grammar?? I don't think we actually have update queries.

A long time ago the possibility of update was being investigated, but it turned out not to be something that existing Query by Method Name solutions had.  I suppose this was a leftover remnant of that.  Thanks for spotting it!